client/rpcserver/rpcserver.go

```diff
@@ -1,20 +1,411 @@
+	out:
+		for {
+			select {
+			case update := <-updates:
+				// Distribute the book to the subscribed clients.
+				log.Tracef("order book update received for " + update.Market)
+			case <-ctx.Done():
+				break out
+			}
+		}
+	}()
+	return m, nil
+}
+
+// add adds a client to the client map, and returns a fresh orderbook.
+func (m *marketSyncer) add(cl *wsClient) *core.OrderBook {
+	m.mtx.Lock()
+	m.clients[cl.cid] = cl
+	m.mtx.Unlock()
+	return m.core.Book(m.dex, m.base, m.quote)
+}
+
+// remove removes a client from the client map. If this is the last client,
+// the market will be "unsynced".
+func (m *marketSyncer) remove(cl *wsClient) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	delete(m.clients, cl.cid)
+	if len(m.clients) == 0 {
+		m.core.Unsync(m.dex, m.base, m.quote)
+	}
+}
+
+// RPCServer is a single-client http and websocket server enabling a JSON
+// interface to the DEX client.
+type RPCServer struct {
+	ctx      context.Context
+	core     clientCore
+	listener net.Listener
+	srv      *http.Server
+	authsha  [32]byte
+	mtx      sync.RWMutex
+	syncers  map[string]*marketSyncer
+	clients  map[int32]*wsClient
+}
+
+// genCertPair generates a key/cert pair to the paths provided.
+func genCertPair(certFile, keyFile string) error {
+	log.Infof("Generating TLS certificates...")
+
+	org := "dcrdex autogenerated cert"
+	validUntil := time.Now().Add(10 * 365 * 24 * time.Hour)
+	cert, key, err := certgen.NewTLSCertPair(elliptic.P521(), org,
+		validUntil, nil)
+	if err != nil {
+		return err
+	}
+
+	// Write cert and key files.
+	if err = ioutil.WriteFile(certFile, cert, 0644); err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(keyFile, key, 0600); err != nil {
+		os.Remove(certFile)
+		return err
+	}
+
+	log.Infof("Done generating TLS certificates")
+	return nil
+}
+
+// handler is the type for functions that handle RPC requests.
+type handler func(s *RPCServer, ws *wsClient, msg *msgjson.Message) *msgjson.Error
+
+// routes maps routes to their handler.
+var routes = map[string]handler{
+	"help":    handleHelp,
+	"version": handleVersion,
+}
+
+// writeJSON marshals the provided interface and writes the bytes to the
+// ResponseWriter. The response code is assumed to be StatusOK.
+func writeJSON(w http.ResponseWriter, thing interface{}) {
+	writeJSONWithStatus(w, thing, http.StatusOK)
+}
+
+// writeJSON writes marshals the provided interface and writes the bytes to the
+// ResponseWriter with the specified response code.
+func writeJSONWithStatus(w http.ResponseWriter, thing interface{}, code int) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(code)
+	encoder := json.NewEncoder(w)
+	if err := encoder.Encode(thing); err != nil {
+		log.Infof("JSON encode error: %v", err)
+	}
+}
+
+// handleJSON handles all https json requests.
+func (s *RPCServer) handleJSON(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Connection", "close")
```

Comment:
From https://tools.ietf.org/html/rfc2616#section-14.10

HTTP/1.1 applications that do not support persistent connections MUST include the "close" connection option in every message.

Reply:
The effect of the close header is clear, but not why we "do not support persistent connections" in the dex client's RPC server.  I see no reason why it couldn't support them.  If it's just a design decision, that's OK and we should document it.

I'm noting that [`http.Request` docs](https://golang.org/pkg/net/http/#Request) state `For server requests, the HTTP server handles this automatically and this field is not needed by Handlers.`, and this is a server handler.  The [client docs](https://golang.org/pkg/net/http/#Client.Do) state that reusing connections may not be possible when `the Body is not both read to EOF and closed`, but this indicates it is the *client's job* to check for read errors and close the body, closing the connection on failure of those actions.

I can imagine a client process that is capable of reusing the connections just as a web browser does.  Even though websocket is better for that, but why deny the possibility of reusing connections for multiple http requests, esp. if the supported routes are different between http and websocket?

Reply:
Anyway, if we are choosing not to support persistent http connections for repeated requests, then let's frame the comment that way and explain why. ` // All http routes are available over websocket too, so do not support persistent http connections.  Inform the user and close the connection when response handling is completed.`  Something like that?  I think a comment is warranted given the http.Request docs stating that the `Close` field is not needed by handlers.

Reply:
Thank you for taking the time to talk with dave about this.  I tried to follow, but I'm not sure, looks ok?

Reply:
Yeah, that comment will work.  I'm still highly skeptical that this is necessary though since (1) the net/http docs suggest otherwise, (2) allowing persistent http connections is a feature, not a bug, that works great with read/write timeouts, and (3) the RPC server is authenticated so abuse is not a concern.